sympy/combinatorics/rewritingsystem.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import print_function, division
+
+from sympy import S
+from sympy.combinatorics.free_groups import FreeGroupElement
+
+class RewritingSystem(object):
+    '''
+    A class implementing rewriting systems for `FpGroup`s.
+
+    References
+    ==========
+    [1] Epstein, D., Holt, D. and Rees, S. (1991).
+        The use of Knuth-Bendix methods to solve the word problem in automatic groups.
+        Journal of Symbolic Computation, 12(4-5), pp.397-414.
+
+    [2] GAP's Manual on its KBMAG package
+        https://www.gap-system.org/Manuals/pkg/kbmag-1.5.3/doc/manual.pdf
+
+    '''
+    def __init__(self, group):
+        self.group = group
+        self.alphabet = group.generators
+        self._is_confluent = None
+
+        # these values are taken from [2]
+        self.maxeqns = 32767 # max rules
+        self.tidyint = 100 # rules before tidying
+
+        # dictionary of reductions
+        self.rules = {}
+        self._init_rules()
+
+    @property
+    def is_confluent(self):
+        '''
+        Return `True` if the system is confluent
+
+        '''
+        if self._is_confluent is None:
+            self._is_confluent = self._check_confluence()
+        return self._is_confluent
+
+    def _init_rules(self):
+        rels = self.group.relators[:]
+        identity = self.group.free_group.identity
+        for r in rels:
+            self.add_rule(r, identity)
+        self._remove_redundancies()
+        return
+
+    def add_rule(self, w1, w2):
+        new_keys = set()
+        if len(self.rules) + 1 > self.maxeqns:
+            return ValueError("Too many rules were defined.")
+
+        if w1 < w2:
+            w1, w2 = w2, w1
+
+        s1, s2 = w1, w2
+
+        # The following is the equivalent of checking
+        # s1 for overlaps with the implicit reductions
+        # {g*g**-1 -> <identity>} and {g**-1*g -> <identity>}
+        # for any generator g without installing the
+        # redundant rules that would result from processing
+        # the overlaps. See [1], Section 3 for details.
+
+        # overlaps on the right
+        while len(s1) - len(s2) > 0:
```

Comment:
If the difference is 1 and no cancelling occurs, then it may become -1. It seems that the rule will not be stored.